src/lib/services/client-metrics/index.js

```diff
@@ -75,30 +78,56 @@ module.exports = class ClientMetricsService {
         });
     }
 
-    async upsertApp(value, clientIp) {
-        try {
-            const app = await this.clientAppStore.getApplication(value.appName);
-            await this.updateRow(value, app);
-        } catch (error) {
-            if (error instanceof NotFoundError) {
-                await this.clientAppStore.insertNewRow(value);
-                await this.eventStore.store({
-                    type: APPLICATION_CREATED,
-                    createdBy: clientIp,
-                    data: value,
-                });
+    async registerClient(data, clientIp) {
+        const value = await clientRegisterSchema.validateAsync(data);
+        value.clientIp = clientIp;
+        this.logger.info(`${JSON.stringify(data)}`);
+        this.seenClients[this.clientKey(value)] = value;
+    }
+
+    clientKey(client) {
+        return `${client.appName}_${client.instanceId}`;
+    }
+
+    async bulkAdd() {
+        if (
+            this &&
```

Comment:
Is this necessary? Given that this is a method on the class, I would image "this" would have to exist in order to call this method?

Reply:
I would too, but had some test issues with test not being defined, though to be honest, I've since changed to binding it with an arrow function, which should make your point valid. I'll drop the check for `this`